Sources/KSCrashRecording/Monitors/KSCrashMonitor_Memory.m

```diff
@@ -0,0 +1,546 @@
+    
+    // check the timestamp, let's say it's valid for the last week
+    // do we really want crash reports older than a week anyway??
+    const uint64_t kUS_in_day = 8.64e+10;
+    const uint64_t kUS_in_week = kUS_in_day * 7;
+    uint64_t now = ksdate_microseconds();
+    if (memory.timestamp <= 0 || memory.timestamp == INT64_MAX || memory.timestamp < now - kUS_in_week) {
+        return;
+    }
+    
+    // check pressure and level are in ranges
+    if (memory.level > KSCrashAppMemoryStateTerminal) {
+        return;
+    }
+    if (memory.pressure > KSCrashAppMemoryStateTerminal) {
+        return;
+    }
+    
+    // check app transition state
+    if (memory.state > KSCrashAppTransitionStateExiting) {
+        return;
+    }
+    
+    // Footprint and remaining should always = limit
+    if (memory.footprint + memory.remaining != memory.limit) {
+        return;
+    }
+    
+    // if we're at max, we likely overflowed or set a negative value,
+    // in any case, we're counting this as a possible error and bailing.
+    if (memory.footprint == UINT64_MAX) {
+        return;
+    }
+    if (memory.remaining == UINT64_MAX) {
+        return;
+    }
+    if (memory.limit == UINT64_MAX) {
+        return;
+    }
+
+    g_previousSessionMemory = memory;
+}
+
+/**
+ Mapping memory to a file on disk. This allows us to simply treat the location
+ in memory as a structure and the kernel will ensure it is on disk. This is also
+ crash resistant.
+ */
+static void ksmemory_map(const char* path)
+{
+    void *ptr = ksfu_mmap(path, sizeof(KSCrash_Memory));
+    if (!ptr) {
+        return;
+    }
+    
+    g_memory = (KSCrash_Memory *)ptr;
+    KSCrashAppMemory *memory = g_memoryTracker.memory;
+    
+    _ks_memory_update(^(KSCrash_Memory *mem) {
+        *mem = (KSCrash_Memory){
+            .footprint = memory.footprint,
+            .remaining = memory.remaining,
+            .pressure = (uint8_t)memory.pressure,
+            .level = (uint8_t)memory.level,
+            .limit = memory.limit,
+            .timestamp = ksdate_microseconds(),
+            .state = KSCrashAppStateTracker.shared.transitionState,
+            .fatal = false,
+        };
+    });
+}
+
+/**
+ What we're doing here is writing a file out that can be reused
+ on restart if the data shows us there was a memory issue.
+ 
+ If an OOM did happen, we'll modify this file
+ (see `kscm_memory_check_for_oom_in_previous_session`),
+ then write it back out using the normal writing procedure to write reports. This
+ leads to the system seeing the report as if it had always been there and will
+ then report an OOM.
+ */
+static void ksmemory_write_possible_oom(void)
+{
+    NSURL *reportURL = kscm_memory_oom_breadcrumb_URL();
+    const char *reportPath = reportURL.path.UTF8String;
+
+    KSMC_NEW_CONTEXT(machineContext);
+    ksmc_getContextForThread(ksthread_self(), machineContext, false);
+    KSStackCursor stackCursor;
```

Comment:
Is this `stackCursor` used?

Reply:
Ah, `kssc_initWithMachineContext` also initializes `machineContext`. Got you, thanks!